packages/core/src/renderer/components/cluster-settings/icon-settings.tsx

```diff
@@ -3,106 +3,113 @@
     }
 
-    const { cluster } = this.props;
-
     try {
       const buf = Buffer.from(await file.arrayBuffer());
 
       cluster.preferences.icon = `data:${file.type};base64,${buf.toString("base64")}`;
     } catch (e) {
-      this.errorText = String(e);
-      this.status = GeneralInputStatus.ERROR;
+      props.showErrorNotification(String(e));
     }
-  }
+  };
 
-  clearIcon() {
-    /**
-     * NOTE: this needs to be `null` rather than `undefined` so that we can
-     * tell the difference between it not being there and being cleared.
-     */
-    this.props.cluster.preferences.icon = null;
-  }
-
-  onUploadClick() {
-    this.element
+  const onUploadClick = () => {
+    element
       .current
       ?.querySelector<HTMLInputElement>("input[type=file]")
       ?.click();
-  }
+  };
 
-  render() {
-    const { entity } = this.props;
+  const save = ([kind, value]: ChangedClusterPreference) => {
+    cluster.preferences[kind] = value;
+  };
 
-    return (
-      <div ref={this.element}>
-        <div className="file-loader flex flex-row items-center">
-          <div className="mr-5">
-            <FilePicker
-              accept="image/*"
-              label={(
-                <Avatar
-                  colorHash={`${entity.getName()}-${entity.metadata.source}`}
-                  title={entity.getName()}
-                  src={entity.spec.icon?.src}
-                  size={53}
-                />
-              )}
-              onOverSizeLimit={OverSizeLimitStyle.FILTER}
-              handler={this.onIconPick}
-            />
-          </div>
-          <MenuActions
-            id={`menu-actions-for-cluster-icon-settings-for-${entity.getId()}`}
-            toolbar={false}
-            autoCloseOnSelect={true}
-            triggerIcon={{ material: "more_horiz" }}
-          >
-            <MenuItem onClick={this.onUploadClick}>
-              Upload Icon
-            </MenuItem>
-            <MenuItem onClick={() => this.clearIcon()} disabled={!this.props.cluster.preferences.icon}>
-              Clear
-            </MenuItem>
-          </MenuActions>
+  return (
+    <div ref={element}>
+      <div className="file-loader flex flex-row items-center">
+        <div className="mr-5">
+          <FilePicker
+            accept="image/*"
+            label={(
+              <Avatar
+                colorHash={`${entity.getName()}-${entity.metadata.source}`}
+                title={entity.getName()}
+                src={entity.spec.icon?.src}
+                size={53}
+              />
+            )}
+            onOverSizeLimit={OverSizeLimitStyle.FILTER}
+            handler={onIconPick}
+          />
         </div>
+        <MenuActions
+          id={menuId}
+          data-testid={menuId}
+          toolbar={false}
+          autoCloseOnSelect={true}
+          triggerIcon={{ material: "more_horiz" }}
+        >
+          <MenuItem onClick={onUploadClick}>
+            Upload Icon
+          </MenuItem>
```

Comment:
Shouldn't this also be separated injectable like you did for "Clear"?

Reply:
Yes, it should. But not doing this because of following tricky part:

```
const onUploadClick = () => {
    element
      .current
      ?.querySelector<HTMLInputElement>("input[type=file]")
      ?.click();
  };
```

Reply:
Do not want to pollute menu items from injectable with a bunch of props such as element reference.